isis/tests/AngleTest.cpp

```diff
@@ -0,0 +1,304 @@
+  EXPECT_NEAR(angle.radians(), Isis::PI, TEST_EPSILON); 
+}
+
+TEST(AngleOperators, QDebug) {
+  // Stream qDebug output to a string and then do a string comparison?
+  Isis::Angle angle(45.0, Isis::Angle::Degrees);
+  QString qdebug_output;
+  QDebug debug(&qdebug_output);
+  debug << angle;
+  EXPECT_STREQ(qdebug_output.toLatin1(), "0.785398 <radians> (45 <degrees>) "); 
+}
+
+
+TEST(AngleOperators, Assignment) {
+  Isis::Angle angle(30.0, Isis::Angle::Degrees );
+  angle = Isis::Angle(45.0, Isis::Angle::Degrees);
+  EXPECT_NEAR(angle.degrees(), 45.0, TEST_EPSILON); 
+}
+
+
+TEST(AngleOperators, AddSubtract) {
+  Isis::Angle angle1(30.0, Isis::Angle::Degrees );
+  Isis::Angle angle2(60.0, Isis::Angle::Degrees );
+  angle1 = angle1 + angle2;  
+  EXPECT_NEAR(angle1.degrees(), 90.0, TEST_EPSILON);
+  angle1 += angle2;
+  EXPECT_NEAR(angle1.degrees(), 150.0, TEST_EPSILON);
+  angle1 -= angle2;
+  EXPECT_NEAR(angle1.degrees(), 90.0, TEST_EPSILON);    
+  angle1 = angle1 - angle2;
+  EXPECT_NEAR(angle1.degrees(), 30.0, TEST_EPSILON);
+}
+
+
+TEST(AngleOperators, MultiplyDivide) {
+  Isis::Angle angle(30., Isis::Angle::Degrees );
+  angle = 2. * angle;
+  EXPECT_NEAR(angle.degrees(), 60.0, TEST_EPSILON);
+  angle *= 2;
+  EXPECT_NEAR(angle.degrees(), 120.0, TEST_EPSILON);
+  angle /= 2;
+  EXPECT_NEAR(angle.degrees(), 60.0, TEST_EPSILON);    
+  angle = angle / 2;
+  EXPECT_NEAR(angle.degrees(), 30.0, TEST_EPSILON);
+}  
+
+
+TEST(AngleOperators, Logical) {
+  Isis::Angle angle1(30.0, Isis::Angle::Degrees );
+  Isis::Angle angle2(45.0, Isis::Angle::Degrees );
+  EXPECT_FALSE(angle1 == angle2);
+  EXPECT_TRUE(Isis::Angle() == Isis::Angle()); 
+  EXPECT_FALSE(Isis::Angle() == angle2); 
+}
+
+
+TEST(AngleOperators, LessThan) {
+  Isis::Angle angle1(30.0, Isis::Angle::Degrees );
+  Isis::Angle angle2(45.0, Isis::Angle::Degrees );
+  EXPECT_TRUE(angle1 < angle2);
+  EXPECT_FALSE(angle2 < angle1);
+  Isis::Angle epsilon(1.1920929e-10, Isis::Angle::Degrees);
+  EXPECT_TRUE(angle1 < angle1 + epsilon); 
+}
+
+// Added good and consistent test for "greater than" 
+TEST(AngleOperators, GreaterThan) {
+  Isis::Angle angle1(30.0, Isis::Angle::Degrees );
+  Isis::Angle angle2(45.0, Isis::Angle::Degrees );
+  EXPECT_TRUE(angle2 > angle1);
+  EXPECT_FALSE(angle1 > angle2); 
+  Isis::Angle epsilon(1.1920929e-10, Isis::Angle::Degrees);
+  EXPECT_TRUE(angle2 > angle2 - epsilon);
+}
+  
+
+TEST(AngleOperators, LessThanOrEqualTo) {
+  Isis::Angle angle1(30.0, Isis::Angle::Degrees );
+  Isis::Angle angle2(45.0, Isis::Angle::Degrees );
+  EXPECT_TRUE(angle1 <= angle1);
+  EXPECT_TRUE(angle1 <= angle2);
+  EXPECT_FALSE(angle2 <= angle1); 
+}
+
+
+TEST(AngleExceptions, LessThanNullAngle){
+  Isis::Angle angle1(30., Isis::Angle::Degrees );
+  
+  try {
+    angle1 < Isis::Angle();
+    FAIL() << "Expected an error";
+   }
+   catch(Isis::IException &e) {
+     EXPECT_EQ(e.toString().toLatin1(), "**PROGRAMMER ERROR** Cannot compare a invalid angles with the < operator." );
```

Comment:
I've had success using the following to test exception messages:

```
EXPECT_THAT(
      e.toString().toStdString(),
      ::testing::HasSubstr("Part of the error message string")
);
```

Reply:
Make sure you add gmock as an include if you use this! `#include "gmock/gmock.h"`

Reply:
Here's the docs on these matchers https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#more-string-assertions

Reply:
And just to document this -- I believe we decided to use something like: 

```
EXPECT_TRUE(Qstring.contains(substring from error message)
```